Various speed optimizations

ryancdotorg opened the pull request:
* Check to see whether compressed data matches already before recompressing if `PRECOMPRESS_OVERWRITE` is enabled
* Don't compress tiny files - controlled by `PRECOMPRESS_MIN_SIZE`, defaults 20
* Use `multiprocessing.pool` to run multiple compression threads

Multiprocessing code based on https://gitlab.com/bryanbrattlof/pelican-htmlmin

kurtmckee commented:
Hey Ryan! Thanks for submitting this.

I'm excited about the code -- it cuts my own site generation down from ten minutes to five minutes, and that's a huge time saver! However, the unit tests don't run anymore. I've identified a few issues that need to be addressed.

* The inclusion of `__init__.py` breaks the unit tests. I'm assuming that you needed to add that file because you've cloned the git repo and are including the repo director itself in your Pelican plugins list, rather than installing it into a virtual environment? I can take the patch, but I can't keep `__init__.py` in the repo.

* After removing `__init_.py` some of the unit tests run, but then one hangs indefinitely. I haven't worked through the details, but it may be related to pyfakefs / multiprocessing.

* It looks like there are no additional unit tests for the new code. I see a few opportunities for the new code to crash. For example, if the plugin opens a .br or .gz file that is corrupt it will immediately crash because no exceptions are caught during decoding. Also, if .br files are encountered but the brotli module isn't installed this will result in a crash as well.

I really like this patch. I'm going to work to fix and update the unit tests before merging.

kurtmckee commented:
I think I've resolved all of the things I mentioned above. These are now merged into the develop branch of this repository and will be included in the next release of pelican-precompress.

In addition to fixing the unit tests I also added support for disabling the minimum file size by setting it to `0`.

I'm not sure how to get GitHub to notice that your commits are now merged into the (now-default) develop branch so I'm manually closing.

Thanks for this commit! I'll put out a new release soon with these changes. Please submit more fixes or improvements as you come up with them! =)

kurtmckee commented:
@ryancdotorg I've pushed a new release. Thanks so much for your work on this!

ryancdotorg commented:
@kurtmckee Thanks for getting it merged, I'm thrilled it saved so much time on your builds, and glad you were able to fix the issues.

 I had an issue with my mail server config that was causing emails from GitHub to be dropped, so unfortunately I didn't see your comments here until just now.

I briefly tried to get the testing framework up and running, but had some problems with that. I see you've now added a `requirements-dev.txt`, which should certainly make that easier in the future.

You're correct that I added the `__init__.py` file to make it work without installing explicitly in the virtualenv. I'm still a bit ignorant of the finer points of python packaging. Is `pip install --editable .` what I'd want to use for that?

kurtmckee commented:
@ryancdotorg, please let me know if you encounter more issues with the test environment. I may have inadvertently tuned it to my own system so if you're not able to run commands like this to get it up and running please let me know!

```shell
$ python -m venv venv
$ source venv/bin/activate
(venv) $ pip install -r requirements-dev.txt
(venv) $ tox
```

This creates a `venv` directory that contains scripts/libraries/etc for Python, and when you want to develop in it just run `source venv/bin/activate` again.

You may be able to skip explicitly installing it; you can add the git repository directory to the list of plugin directories in Pelican, and then add the filename as a string for Pelican to hunt for and load. For example, I uninstalled the plugin from my virtual environment and then tested that this configuration works as desired.

```python
# pelicanconf.py
PLUGIN_PATHS = ['/home/user/git-repos/pelican-precompress']
PLUGINS = ['pelican_precompress']
```

Pelican will do the work of hunting down the right file from there.

I'm not familiar with the `--editable` option but I also tested running these commands:

```shell
$ # If you use a virtual environment for Pelican, activate it first.
$ # Otherwise, just run these commands.
$ cd /home/user/git-repos/pelican-precompress
$ python setup.py install
```

That'll install the package to your Pelican virtual environment or to your Python global library directory, depending on if you activated a virtual environment beforehand.

kurtmckee commented:
@ryancdotorg please let me know if that worked, or if you have follow-up questions. =)
